File: python/dssatutils/weather_terraclimate.py
# Weather source: TerraClimate monthly climate, disaggregated to daily DSSAT input.
#
# TerraClimate (Abatzoglou et al. 2018) is a MONTHLY product (~4 km), so it cannot
# be written straight to a daily DSSAT .WTH. This backend expands each month into
# continuous daily records: Tmax/Tmin/SRAD/wind are held constant at the monthly
# value and the monthly precipitation TOTAL is spread uniformly across the days of
# the month. The result is a runnable but deliberately smooth weather file —
# intended for screening / climatology only, NOT for analyses that depend on
# day-to-day variability (rainfall intensity, dry spells, heat-stress days).

import calendar

import numpy as np
import pandas as pd

from .weather_gridded_common import (
    extract_netcdf_series, find_nc_file, write_wth,
)

# DSSAT var -> (filename tokens, NetCDF var aliases, unit-conversion kind)
_VARS = {
    "TMAX": (["tmmx", "tmax"], ["tmmx", "tmax"], "temp"),
    "TMIN": (["tmmn", "tmin"], ["tmmn", "tmin"], "temp"),
    "RAIN": (["ppt", "precip", "pr"], ["ppt", "precip", "pr"], "rain"),
    "SRAD": (["srad", "rsds"], ["srad", "rsds"], "srad"),
    "WIND": (["ws", "wind"], ["ws", "wind"], "wind"),
}


def _coords(shapefile, id_col, lat_col, lon_col):
    pts = shapefile.copy()
    if hasattr(pts, "geometry"):
        pts = pts.to_crs("EPSG:4326")
        lats = pts.geometry.y.values.astype(float)
        lons = pts.geometry.x.values.astype(float)
    else:
        lats = pts[lat_col].astype(float).values
        lons = pts[lon_col].astype(float).values
    ids = [str(r[id_col]) for _, r in pts.iterrows()]
    return ids, lats, lons


def _expand_month_to_daily(monthly_by_var):
    """monthly_by_var: {DSSAT_VAR: {YYYYDOY(month-start): value}} -> daily DataFrame.

    Returns continuous daily rows for every month for which TMAX and TMIN exist.
    """
    months = sorted(set(monthly_by_var.get("TMAX", {})) & set(monthly_by_var.get("TMIN", {})))
    rows = []
    for code in months:
        year = int(code[:4])
        month = (pd.Timestamp(f"{year}-01-01") + pd.Timedelta(days=int(code[4:]) - 1)).month
        ndays = calendar.monthrange(year, month)[1]
        tmax = monthly_by_var["TMAX"][code]
        tmin = monthly_by_var["TMIN"][code]
        srad = monthly_by_var.get("SRAD", {}).get(code, -99.0)
        wind = monthly_by_var.get("WIND", {}).get(code, -99.0)
        ppt_total = monthly_by_var.get("RAIN", {}).get(code, np.nan)
        rain_daily = (ppt_total / ndays) if np.isfinite(ppt_total) else -99.0
        for dom in range(1, ndays + 1):
            d = pd.Timestamp(year=year, month=month, day=dom)
            rows.append({
                "DATE": f"{d.year}{d.dayofyear:03d}", "YEAR": d.year, "MM": d.month,
                "SRAD": srad, "TMAX": tmax, "TMIN": tmin, "RAIN": rain_daily,
                "TDEW": -99.0, "RH2M": -99.0, "WIND": wind,
            })
    return pd.DataFrame(rows)


def process_weather_terraclimate(
    shapefile, start_year, end_year, output_dir,
    id_col, lat_col, lon_col, n_cores, log_file,
    terraclimate_nc_dir: str,
) -> None:
    """Write screening-grade daily DSSAT .WTH from TerraClimate monthly NetCDF.

    TerraClimate is monthly; this backend disaggregates each month to continuous
    daily records (constant Tmax/Tmin/SRAD/wind, monthly precip spread evenly).
    Use for screening/climatology only — it carries no day-to-day variability.
    """
    import os
    if not terraclimate_nc_dir or not os.path.isdir(terraclimate_nc_dir):
        raise FileNotFoundError(
            f"TerraClimate needs a local NetCDF directory: {terraclimate_nc_dir}")
    os.makedirs(output_dir, exist_ok=True)
    print("WARNING: TerraClimate is monthly; disaggregated to daily for screening/climatology only.")
    print(f"--- Starting TerraClimate Processing (Years: {start_year}-{end_year}) ---")

    ids, lats, lons = _coords(shapefile, id_col, lat_col, lon_col)

    per_var = {}
    for dssat_var, (tokens, aliases, kind) in _VARS.items():
        path = find_nc_file(terraclimate_nc_dir, tokens)
        if not path:
            if dssat_var in ("TMAX", "TMIN", "RAIN"):
                raise FileNotFoundError(
                    f"TerraClimate required variable {dssat_var} not found in {terraclimate_nc_dir}")
            continue
        per_var[dssat_var] = extract_netcdf_series(
            path, aliases, ids, lats, lons, int(start_year), int(end_year), kind)

    written = 0
    for pid, lat, lon in zip(ids, lats, lons):
        try:
            monthly_by_var = {v: per_var.get(v, {}).get(pid, {}) for v in _VARS}
            frame = _expand_month_to_daily(monthly_by_var)
            if frame.empty:
                raise ValueError("No overlapping monthly TMAX/TMIN extracted for this point.")
            write_wth(frame, pid, lat, lon, output_dir, "TerraClimate monthly->daily", "TCLM")
            written += 1
        except Exception as exc:  # noqa: BLE001
            msg = f"\n--- ERROR ---\nTerraClimate point {pid} ({lat:.3f},{lon:.3f}): {exc}\n"
            print(msg)
            with open(log_file, "a") as lf:
                lf.write(msg)
    print(f"\nTerraClimate processing complete: {written} point(s) written.\n")

File: R/weather_terraclimate.R
# Weather source: TerraClimate monthly climate, disaggregated to daily DSSAT input.
#
# TerraClimate is MONTHLY (~4 km); it cannot be written straight to a daily .WTH.
# This backend expands each month into continuous daily records: Tmax/Tmin/SRAD/
# wind are held at the monthly value and the monthly precip TOTAL is spread evenly
# across the days of the month. Runnable but deliberately smooth -- screening /
# climatology only, NOT for day-to-day variability (intensity, dry spells, stress).

# DSSAT var -> (filename tokens, unit-conversion kind, required?)
TERRACLIMATE_WEATHER_VARS <- list(
  TMAX = list(tokens = c("tmmx", "tmax"), kind = "temp", required = TRUE),
  TMIN = list(tokens = c("tmmn", "tmin"), kind = "temp", required = TRUE),
  RAIN = list(tokens = c("ppt", "precip", "pr"), kind = "rain", required = TRUE),
  SRAD = list(tokens = c("srad", "rsds"), kind = "srad", required = FALSE),
  WIND = list(tokens = c("wind", "ws"), kind = "wind", required = FALSE)
)

.terraclimate_expand_daily <- function(monthly_by_var, pid) {
  tmax_m <- monthly_by_var[["TMAX"]][[pid]]
  tmin_m <- monthly_by_var[["TMIN"]][[pid]]
  if (is.null(tmax_m) || is.null(tmin_m)) return(NULL)
  codes <- intersect(names(tmax_m), names(tmin_m))
  if (!length(codes)) return(NULL)
  codes <- sort(codes)
  get1 <- function(v, code) {
    s <- monthly_by_var[[v]][[pid]]
    if (is.null(s) || is.na(s[code])) NA_real_ else as.numeric(s[code])
  }
  frames <- list()
  for (code in codes) {
    year <- as.integer(substr(code, 1, 4))
    doy <- as.integer(substr(code, 5, 7))
    month <- as.integer(format(as.Date(sprintf("%d-01-01", year)) + (doy - 1), "%m"))
    ndays <- as.integer(format(
      as.Date(sprintf("%d-%02d-01", ifelse(month == 12, year + 1, year),
                      ifelse(month == 12, 1, month + 1))) - 1, "%d"))
    srad <- get1("SRAD", code); wind <- get1("WIND", code)
    ppt <- get1("RAIN", code)
    rain_daily <- if (is.finite(ppt)) ppt / ndays else -99
    days <- as.Date(sprintf("%d-%02d-01", year, month)) + (seq_len(ndays) - 1)
    frames[[code]] <- data.frame(
      DATE = sprintf("%d%03d", year, as.integer(format(days, "%j"))),
      YEAR = year, MM = month,
      SRAD = ifelse(is.finite(srad), srad, -99),
      TMAX = as.numeric(tmax_m[code]), TMIN = as.numeric(tmin_m[code]),
      RAIN = rain_daily, TDEW = -99, RH2M = -99,
      WIND = ifelse(is.finite(wind), wind, -99))
  }
  do.call(rbind, frames)
}

process_weather_terraclimate <- function(shapefile, start_year, end_year, output_dir,
                                         id_col, lat_col, lon_col, n_cores, log_file,
                                         terraclimate_nc_dir) {
  if (!nzchar(terraclimate_nc_dir) || !dir.exists(terraclimate_nc_dir))
    stop(sprintf("TerraClimate needs a local NetCDF directory: %s", terraclimate_nc_dir))
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  warning("TerraClimate is monthly; disaggregated to daily for screening/climatology only.")
  message(sprintf("--- Starting TerraClimate Processing (Years: %d-%d) ---", start_year, end_year))

  pts <- sf::st_transform(shapefile, 4326)
  ids <- as.character(sf::st_drop_geometry(pts)[[id_col]])
  pts_vect <- terra::vect(pts)
  xy <- sf::st_coordinates(pts); lats <- xy[, 2]; lons <- xy[, 1]

  per_var <- list()
  for (v in names(TERRACLIMATE_WEATHER_VARS)) {
    spec <- TERRACLIMATE_WEATHER_VARS[[v]]
    path <- weather_find_nc_file(terraclimate_nc_dir, spec$tokens)
    if (is.na(path)) {
      if (isTRUE(spec$required))
        stop(sprintf("TerraClimate required variable %s not found in %s", v, terraclimate_nc_dir))
      next
    }
    per_var[[v]] <- weather_extract_netcdf_series(path, ids, pts_vect,
                                                  start_year, end_year, spec$kind)
  }

  written <- 0
  for (k in seq_along(ids)) {
    pid <- ids[k]
    tryCatch({
      df <- .terraclimate_expand_daily(per_var, pid)
      if (is.null(df) || !nrow(df)) stop("No overlapping monthly TMAX/TMIN extracted for this point.")
      weather_write_wth(df, pid, lats[k], lons[k], output_dir, "TerraClimate monthly->daily", "TCLM")
      written <- written + 1
    }, error = function(e) {
      msg <- sprintf("\n--- ERROR ---\nTerraClimate point %s: %s\n", pid, conditionMessage(e))
      cat(msg); write(msg, file = log_file, append = TRUE)
    })
  }
  message(sprintf("\nTerraClimate processing complete: %d point(s) written.\n", written))
}
